source/FormationPattern.h

```diff
@@ -0,0 +1,95 @@
+/* FormationPattern.h
+Copyright (c) 2019-2022 by Peter van der Meer
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef FORMATION_PATTERN_H_
+#define FORMATION_PATTERN_H_
+
+#include "DataNode.h"
+#include "Point.h"
+
+#include <string>
+#include <vector>
+
+class Body;
+
+
+
+// Class that handles the loading and position calculations for a pattern that
+// can be used for ships flying in formation.
+// This class only deals with calculating the positions that exist in a formation
+// pattern, the actual assignment of ships to positions is handled outside this class.
+class FormationPattern {
+public:
+	// Iterator that provides sequential access to all formation positions.
+	class PositionIterator
+	{
+	public:
+		PositionIterator(const FormationPattern &pattern);
+		PositionIterator() = delete;
+
+		// Iterator traits
+		using iterator_category = std::input_iterator_tag;
+		using value_type = Point;
+		using difference_type = void;
+		using pointer = const Point *;
+		using reference = Point &;
+
+		// A subset of the default input_iterator operations. Limiting to
+		// only a subset, since not all operations are used in-game.
+		const Point *operator->();
+		const Point &operator*();
+		PositionIterator &operator++();
+
+	private:
+		void MoveToValidPosition();
+
+	private:
+		// The pattern for which we are calculating positions.
+		const FormationPattern &pattern;
+		// The location in the pattern.
+		unsigned int slot = 0;
+		// Currently calculated Point.
+		Point currentPoint;
+	};
+
+
+public:
+	// Returns the name of this pattern.
+	const std::string &Name() const;
+	void SetName(const std::string &name);
+
+	// Load formation from a datafile.
+	void Load(const DataNode &node);
+
+	// Get an iterator to iterate over the formation positions in this pattern.
+	PositionIterator begin() const;
+
+
+private:
+	// Retrieve properties like the number of positions.
+	unsigned int Slots() const;
+
+	// Calculate a position based on the formation definition.
+	Point Position(unsigned int slotNr) const;
```

Comment:
Any reason this doesn't return a reference?

Reply:
Historical reasons; the previous more complex versions of this code calculated the positions on the fly, so there were no references to return.

The current version of the code only works with positions/Points, so we could indeed return references here.

Reply:
I updated the code to completely remove this function. It now uses direct references to the positions in the formation, or to the default location (Point(0, 0)) when no valid position is available.